chore(pdi): add data migration for pdi to pdi2

hazzadous opened the pull request:
## Changes

This PR adds a new async migration to populate the new `person_distinct_id2` table.

The new table allows us to figure out the `person_id` for a given distinct_id much faster and using less memory. For example, the new table allows us to get the teams distinct_ids 20x faster with 20M users.

## How did you test this code?

See unit tests

Commit:
chore(pdi): add data migration for pdi to pdi2

This adds an async migration to copy the latest non-deleted
`(team_id, person_id, distinct_id)` tuples from `pdi` to `pdi2`.

Note that this has already be performed for tead_id = 2 on posthog,
cloud so we ensure we're maintaining parity with however this
migration was performed. I've done this by running:

```
SELECT * FROM <old_query>
FULL JOIN <new_query> new
    ON old.distinct_id = new.distinct_id
WHERE old.person_id <> new.person_id
```

specifically for team_id = 2.

macobo commented on posthog/async_migrations/migrations/0002_fill_pdi2_from_pdi.py:
```diff
@@ -0,0 +1,140 @@
+                    is_deleted, 
+                    version
+                )
+                
```
Nit: Confusing blank line :)

hazzadous replied:
I've put the SQL through a formatter now

macobo commented on posthog/async_migrations/migrations/0002_fill_pdi2_from_pdi.py:
```diff
@@ -0,0 +1,140 @@
+                    0 as is_deleted,
+                    0 as version
+
+                FROM
```
Note I think think we might run out of memory for this on cloud.

hazzadous replied:
I haven't looked at this properly yet, but yes it doesn't like such a large query

hazzadous replied:
I've added a comment regarding oom considerations

macobo commented on posthog/async_migrations/migrations/0002_fill_pdi2_from_pdi.py:
```diff
@@ -0,0 +1,140 @@
+                            
+                            Also, use the latest timestamp.
+                            
+                            Note that this query differs from the existing 
```
Q: Why does it differ? I'm not confident this is actually correct. The scenario I'm considering is that:
1. Person gets merged
2. We emit rows with equal timestamps.

Note that there have been man-months spent on proving and making the previous query correct by Yakko/Tiina.

I posit it's easier and safer to just do the same thing we do in queries and not reinvent the wheel - reducing risk here is important.

hazzadous replied:
This is the query from https://github.com/PostHog/posthog/issues/7663

The comment is in reference to being different to https://github.com/PostHog/posthog/blob/78a787bc3bb2d6eaefb8efd2114453e21e58ac33/ee/clickhouse/sql/person.py#L255

E.g. it does not have the further wrapping. If I add a comment as to the source of this and explicitly the difference would this be enough?

hazzadous replied:
I've added a comment regarding oom considerations

hazzadous replied:
Differences to https://github.com/PostHog/posthog/issues/7663:

 1. doesn't filter by team_id
 2. groups by team_id

macobo replied:
Ah! Sorry for getting this wrong!

macobo commented on posthog/async_migrations/migrations/0002_fill_pdi2_from_pdi.py:
```diff
@@ -0,0 +1,140 @@
+            this migration)
+         2. read all non-deleted (team_id, distinct_id, person_id) rows from pdi
+            into pdi2 (this migration)
+         3. update app to always use pdi2 (separate PR)
```
We can't just throw away the old table yet.

This is already somewhat handled. We need to make the migration use similar toggles to what james did [here](https://github.com/PostHog/posthog/pull/7662).

macobo replied:
Note that the documentation here should ideally be evergreen - mentioning separate PRs is noise for any reader.

macobo commented on posthog/async_migrations/migrations/0002_fill_pdi2_from_pdi.py:
```diff
@@ -0,0 +1,140 @@
+        additional rows should not be an issue as we're adding these with
+        version = 0.
+
+        The migration strategy:
```
One of the open questions unanswered from the linked issue: When can we delete the old table and code handling it?

macobo commented on posthog/async_migrations/migrations/0002_fill_pdi2_from_pdi.py:
```diff
@@ -0,0 +1,140 @@
+        within postgres, then propagated to clickhouse on successful commit via
+        kafka.
+
+        NOTE: on cloud we have already migrated team_id = 2, but adding
```
Incorrect (we also have another team under it) and not really relevant context to be saved forever in this migration

macobo commented on posthog/async_migrations/migrations/0002_fill_pdi2_from_pdi.py:
```diff
@@ -0,0 +1,140 @@
+        Schema change to migrate the data from the old person_distinct_id table
+        to the new person_distinct_id2 table. This is required as the existing
+        person_distinct_id does not use a strictly increasing id for
+        identifying the latest (distinct_id, person_id) pairing.
```
This reason is wrong - it was possible to correctly query it, just it was really slow for our largest users. The versioning logic is just a cherry on top to simplify things.

hazzadous replied:
So is the assumption that you can't undelete a person and thus if there is a delete we know it must be the last row? It wasn't clear to me if this was the case but yup if that is the case I guess we don't care so much.

macobo commented on posthog/async_migrations/migrations/0002_fill_pdi2_from_pdi.py:
```diff
@@ -0,0 +1,140 @@
+        might want to perform aggregation over data, but `ReplacingMergeTree` is
+        a better fix where we just want the latest data.
+
+        The new schema includes a `version` column which should be strictly
```
Replace `should` with `is`. No reason to be unconfident here.

macobo commented on posthog/async_migrations/migrations/0002_fill_pdi2_from_pdi.py:
```diff
@@ -0,0 +1,140 @@
+    """
+
+    posthog_min_version = "1.31.0"
+    posthog_max_version = "1.32.0"
```
Not sure this is correct - will need to verify with the async' migrations creator.

macobo commented:
Really solid start, really appreciate the in-depth documentation. Had some concerns though, might take a look into collaborating with @yakkomajuri or @tiina303 to get this live over the next weeks. :)

hazzadous commented:
> Really solid start, really appreciate the in-depth documentation. Had some concerns though, might take a look into collaborating with @yakkomajuri or @tiina303 to get this live over the next weeks. :)

Yup for sure, this PR should hopefully be a talking point if not the eventual approach

macobo commented:
Summary from chatting with @yakkomajuri, how to get this live on prod:
- [ ] We should implement a new constance flag that will determine whether to be reading from the new table and whether the old table (defaults to False for existing installs). Set this in side_effects
    - [ ] Need to bring support for that flag into plugin-server
    - [ ] Flag should be set to True in new installs (e.g. by changing one of the old ee/clickhouse migrations to set the flag)
- [ ] The migration should have a is_required that depends on the constance flag - skip if all needed is already done!
- [ ] We should run the query team-by-team. Can do so by changing the semantics of migration.operations array - e.g. making it a cached_property in this case. This works since this will be run (manually) after code has been updated.
- [ ] Precheck/healthcheck: Cannibalize the 0002 migration, extract common wisdom
- [ ] To eventually delete the migration, set max version we'll support. E.g. 1.35.0 seems reasonable to remove old code at!

macobo committed:
Rename migration

macobo committed:
Skip 0003_fill_person_distinct_id2 on fresh installs

macobo committed:
Clarify version requirements

macobo committed:
Run async migrations using a while-loop instead of tail recursion

Python has a stack limit of 1000, which we might easily run into for
0003 migration

macobo committed:
Use built-in progress tracking

macobo committed:
Make description fit into database 400 char limit

macobo committed:
Add correctness test for new async migration

macobo committed:
Migrate person_distinct_id2 team-by-team

macobo committed:
Remove dead code

macobo committed:
Update migration notes

macobo committed:
Fix foss tests

tiina303 commented on posthog/async_migrations/migrations/0003_fill_person_distinct_id2.py:
```diff
@@ -0,0 +1,90 @@
+
+    1. write to both pdi and pdi2 any new updates (done prior to this migration)
+    2. insert all non-deleted (team_id, distinct_id, person_id) rows from pdi into pdi2 (this migration)
+    3. Once migration has run, we only read/write from/to pdi2.
```
In order for reverts to be easier if anything goes wrong perhaps we could write to both pdi & pdi2 for a little while?

macobo replied:
What is the scenario you are worried about?

For cloud, turning off writes is handled in a separate PR.
For self-hosted, what would be the point? We'd significantly slow ourselves down for it.
If reasoning is failed migrations, see the linked plugin server pr. We only stop writes if this migration is marked a success

tiina303 commented on posthog/async_migrations/migrations/0003_fill_person_distinct_id2.py:
```diff
@@ -0,0 +1,90 @@
+    posthog_max_version = "1.34.0"
+
+    def is_required(self):
+        rows = sync_execute(
+            """
+            SELECT comment
+            FROM system.columns
+            WHERE database = %(database)s AND table = 'person_distinct_id' AND name = 'distinct_id'
+        """,
+            {"database": CLICKHOUSE_DATABASE},
+        )
+
+        return len(rows) > 0 and rows[0][0] != "skip_0003_fill_person_distinct_id2"
```
we can just `return True` if we mark the migration completed during new installs as I suggested above

tiina303 commented on posthog/async_migrations/runner.py:
```diff
@@ -77,23 +79,32 @@ def start_async_migration(migration_name: str, ignore_posthog_version=False) ->
+
 
+def run_async_migration_operations(migration_name: str, migration_instance: Optional[AsyncMigration] = None) -> bool:
+    while True:
+        run_next, success = run_async_migration_next_op(migration_name, migration_instance)
+        if not run_next:
+            return success
```
👍  adding this function
Consider (so `run _next_op` could just return success/failure which would be easier to read imho)
```suggestion
    while (not migration_instance or migration_instance.status !=  MigrationStatus.CompletedSuccessfully):
        if not run_async_migration_next_op(migration_name, migration_instance):
          return False
    return True
```

macobo replied:
The proposed refactoring does not work for handling failures. Will think on the idea tomorrow.

tiina303 replied:
If there's a failure we'd `return False` (if `run_async_migration_next_op` failed, it doesn't return `run_next` only `success` based on my proposal) on line 88 in my proposal, or did I misunderstand what you meant by handling failures?

fuziontech replied:
I think the original here is very easy to read if you want a second opinion here

tiina303 reviewed (approved):
lgtm, added some comments to consider

tiina303 commented on ee/clickhouse/migrations/0003_person.py:
```diff
@@ -1,5 +1,18 @@
+operations = [
+    migrations.RunSQL(PERSONS_TABLE_SQL()),
+    migrations.RunSQL(PERSONS_DISTINCT_ID_TABLE_SQL()),
+    # :TRICKY: This is only run on new installations, we use this to know to skip
+    # posthog/async_migrations/migrations/0003_fill_person_distinct_id2.py
+    # We would use table comments but can't due to clickhouse version limitations
+    migrations.RunSQL(
+        f"""
+        ALTER TABLE person_distinct_id ON CLUSTER {CLICKHOUSE_CLUSTER}
+        COMMENT COLUMN distinct_id 'skip_0003_fill_person_distinct_id2'
+    """
+    ),
```
Potentially a separate MigrationState `CompletedDuringInitialInstall` or something similar.

macobo replied:
I disagree with the proposal - that would make is_required effectively meaningless for all migrations. That's an option but let's prioritize consistency within this lr.

macobo commented on posthog/async_migrations/migrations/0003_fill_person_distinct_id2.py:
```diff
@@ -0,0 +1,90 @@
+    depends_on = "0002_events_sample_by"
+
+    # After releasing this version we can remove code related to `person_distinct_id` table
+    posthog_max_version = "1.34.0"
```
Leaving a task rotting in a task queue to clean it up would imo just cause issues.

There's a real maintenance cost to maintaining the old read-write paths that might prevent other upcoming work so 1.34.0 is the aggressive deadline which we do want to enforce here. Also chatted with yakko about this :)

fuziontech commented on posthog/async_migrations/migrations/0003_fill_person_distinct_id2.py:
```diff
@@ -0,0 +1,90 @@
+            resumable=True,
+        )
+
+    @cached_property
```
solid use of memoization 🧠

fuziontech commented on posthog/async_migrations/runner.py:
```diff
@@ -77,23 +79,32 @@ def start_async_migration(migration_name: str, ignore_posthog_version=False) ->
-    If `run_all=True`, we run through all operations recursively, else we run one and return
+    We run the next operation of the migration which needs attention
+
+    Returns (run_next, success)
```
Really solid refactor. Very clear what states we can get into with `Returns (run_next, success)`

fuziontech reviewed (approved):
This looks great to me.

A few very pedantic suggestions.

:shipit:
